File: src/main/java/com/goudourasv/data/institutions/InstitutionsRepository.java
package com.goudourasv.data.institutions;


import com.goudourasv.domain.institutions.Institution;
import com.goudourasv.http.institutions.dto.InstitutionCreate;
import com.goudourasv.http.institutions.dto.InstitutionUpdate;

import javax.enterprise.context.ApplicationScoped;
import javax.persistence.EntityManager;
import javax.persistence.Query;
import javax.ws.rs.NotFoundException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static com.goudourasv.data.institutions.InstitutionsMapper.toInstitution;
import static com.goudourasv.data.institutions.InstitutionsMapper.toInstitutions;

@ApplicationScoped
public class InstitutionsRepository {
    private final EntityManager entityManager;

    public InstitutionsRepository(EntityManager entityManager) {
        this.entityManager = entityManager;
    }

    public List<Institution> getInstitutions(String country, String city) {
        String sqlQuery = "SELECT * FROM institutions";
        if (country != null || city != null) {
            sqlQuery += " WHERE ";
        }

        Map<String, Object> parametersMap = new HashMap<>();

        boolean isFirst = true;
        if (country != null) {
            if (isFirst) {
                sqlQuery += "country = :country";
                isFirst = false;
            } else {
                sqlQuery += " AND country = :country";
            }
            parametersMap.put("country", country);
        }
        if (city != null) {
            if (isFirst) {
                sqlQuery += "city = :city";
                isFirst = false;
            } else {
                sqlQuery += " AND city = :city";
            }
            parametersMap.put("city", city);
        }

        Query query = entityManager.createNativeQuery(sqlQuery, InstitutionEntity.class);
        for (String key : parametersMap.keySet()) {
            query.setParameter(key, parametersMap.get(key));
        }

        @SuppressWarnings("unchecked")
        List<InstitutionEntity> institutionEntities = query.getResultList();
        List<Institution> institutions = toInstitutions(institutionEntities);

        return institutions;
    }

    public Institution getSpecificInstitution(UUID id) {
        InstitutionEntity institutionEntity = entityManager.find(InstitutionEntity.class, id);
        if(institutionEntity == null){
            return null;
        }
        Institution institution = toInstitution(institutionEntity);

        return institution;
    }


    public Institution createInstitution(InstitutionCreate institutionCreate) {
        InstitutionEntity institutionEntity = new InstitutionEntity();
        institutionEntity.setName(institutionCreate.getName());
        institutionEntity.setCity(institutionCreate.getCity());
        institutionEntity.setCountry(institutionCreate.getCountry());

        entityManager.persist(institutionEntity);
        entityManager.flush();

        Institution institution = toInstitution(institutionEntity);
        return institution;
    }


    public boolean deleteSpecificInstitution(UUID id) {
        InstitutionEntity institutionEntity = entityManager.getReference(InstitutionEntity.class, id);
        if (institutionEntity == null) {
            return false;
        }
        entityManager.remove(institutionEntity);
        return true;

    }

    public Institution replaceInstitution(InstitutionCreate institutionCreate, UUID id) {
        InstitutionEntity institutionEntity = entityManager.getReference(InstitutionEntity.class, id);
        institutionEntity.setId(id);
        institutionEntity.setName(institutionCreate.getName());
        institutionEntity.setCountry(institutionCreate.getCountry());
        institutionEntity.setCity(institutionCreate.getCity());
        try {
            entityManager.merge(institutionEntity);
        } catch (Exception ex) {
            throw new NotFoundException("Institution with id: " + id + "doesn't exist");
        }

        Institution institution = toInstitution(institutionEntity);
        return institution;
    }

    public Institution partiallyUpdateInstitution(InstitutionUpdate institutionUpdate, UUID id) {
        InstitutionEntity institutionEntity = entityManager.getReference(InstitutionEntity.class, id);

        if (institutionUpdate.getName() != null) {
            String newInstitutionName = institutionUpdate.getName();
            institutionEntity.setName(newInstitutionName);
        }
        if (institutionUpdate.getCountry() != null) {
            String newInstitutionCountry = institutionUpdate.getCountry();
            institutionEntity.setCountry(newInstitutionCountry);
        }
        if (institutionUpdate.getCity() != null) {
            String newInstitutionCity = institutionUpdate.getCity();
            institutionEntity.setCity(newInstitutionCity);
        }
        entityManager.merge(institutionEntity);
        entityManager.flush();

        Institution institution = toInstitution(institutionEntity);
        return institution;
    }
}

File: src/main/java/com/goudourasv/data/institutions/InstitutionsRepository.kt
package com.goudourasv.data.institutions

import com.goudourasv.domain.institutions.Institution
import com.goudourasv.http.institutions.dto.InstitutionCreate
import com.goudourasv.http.institutions.dto.InstitutionUpdate
import java.util.*
import javax.enterprise.context.ApplicationScoped
import javax.persistence.EntityManager
import javax.persistence.Query
import javax.ws.rs.NotFoundException

@ApplicationScoped
class InstitutionsRepository(private val entityManager: EntityManager) {
    //TODO Refactor this ugly function
    fun getInstitutions(country: String?, city: String?): List<Institution> {
        var sqlQuery = " SELECT * FROM institutions"
        if (country != null || city != null) {
            sqlQuery += " WHERE "
        }

        val parametersMap = mutableMapOf<String, Any>()

        var isFirst = true
        if (country != null) {
            if (isFirst) {
                sqlQuery += "country = :country"
                isFirst = false
            } else {
                sqlQuery += " AND country = :country"
            }
            parametersMap["country"] = country
        }
        if (city != null) {
            sqlQuery += if (isFirst) {
                "city = :city"

            } else {
                " AND city = :city"
            }
            parametersMap["city"] = city
        }
        val query: Query = entityManager.createNativeQuery(sqlQuery, InstitutionEntity::class.java)
        for (key in parametersMap.keys) {
            query.setParameter(key, parametersMap[key])
        }
        @Suppress("UNCHECKED_CAST")
        val institutionEntities = query.resultList as List<InstitutionEntity>
        return institutionEntities.toInstitutions()
    }

    fun getSpecificInstitution(id: UUID): Institution? {
        val institutionEntity = entityManager.find(InstitutionEntity::class.java, id) ?: return null
        return institutionEntity.toInstitution()
    }

    fun createInstitution(institutionCreate: InstitutionCreate): Institution {
        val institutionEntity = InstitutionEntity(
            name = institutionCreate.name,
            city = institutionCreate.city,
            country = institutionCreate.country
        )
        entityManager.persist(institutionEntity)
        entityManager.flush()
        return institutionEntity.toInstitution()
    }

    //Handling the association before deleting element from non-owner
    fun deleteSpecificInstitution(id: UUID): Boolean {
        val institutionEntity = entityManager.find(InstitutionEntity::class.java, id) ?: return false
        institutionEntity.instructorEntities?.forEach { instructorEntity ->
            instructorEntity.institutionEntities?.remove(
                institutionEntity
            )
        }
        entityManager.remove(institutionEntity)
        return true
    }

    fun replaceInstitution(institutionCreate: InstitutionCreate, id: UUID): Institution {
        val institutionEntity = entityManager.find(InstitutionEntity::class.java, id)
            ?: throw NotFoundException("Institution with id: $id doesn't exist")
        institutionEntity.name = institutionCreate.name
        institutionEntity.city = institutionCreate.city
        institutionEntity.country = institutionCreate.country
        entityManager.merge(institutionEntity)
        entityManager.flush()


        return institutionEntity.toInstitution()
    }

    fun partiallyUpdateInstitution(institutionUpdate: InstitutionUpdate, id: UUID): Institution {
        val institutionEntity = entityManager.getReference(InstitutionEntity::class.java, id)

        if (institutionUpdate.name != null) {
            institutionEntity.name = institutionUpdate.name
        }
        if (institutionUpdate.city != null) {
            institutionEntity.city = institutionUpdate.city
        }
        if (institutionUpdate.country != null) {
            institutionEntity.country = institutionUpdate.country
        }

        entityManager.merge(institutionEntity)
        entityManager.flush()

        return institutionEntity.toInstitution()

    }


}